Microseconds

dannyroberts opened the pull request:
I'm kind of piloting the overall structure of what I'm going to do with timezones by doing the smaller job of adding μs support to datetimes in forms and cases (going forward, on new submissions).

dannyroberts committed:
fix case and form tests

dannyroberts committed:
fix receiverwrapper tests

after timezone changes

dannyroberts committed:
fix pact tests

dannyroberts committed:
add TransitionalExactDateTimeProperty

dannyroberts committed:
make cases and form save usecs

dannyroberts committed:
fix casexml.apps.phone tests

dimagimon commented on corehq/ext/jsonobject.py:
```diff
@@ -0,0 +1,69 @@
+    r'([zZ]|([\+-])([01]\d|2[0-3])\D?([0-5]\d)?)?$'
+)
+
+class TransDateTimeMeta(object):
```
E302 expected 2 blank lines, found 1

dannyroberts commented:
Going to put this on staging; partly just PRing to run all the tests

dannyroberts committed:
update dimagi utils to make tests pass

dannyroberts committed:
fix callcenter tests

dannyroberts committed:
use new USecDateTimeProperty in commtrack

replaced all usages of DateTimeProperty
in apps: commtrack, locations, products, programs

dannyroberts committed:
Merge branch 'master' into microseconds

dannyroberts commented:
Yay! Tests passing. I think now's a good time to start reviewing, and I'll throw it up on staging.

@czue @NoahCarnahan

NoahCarnahan commented:
When should a class inherit from `USecDocument` instead of `Document`?

snopoke commented on corehq/apps/callcenter/fixturegenerators.py:
```diff
@@ -13,17 +15,26 @@ def should_sync(domain, last_sync, utcnow=None):
+
+    if not _assert(last_sync_utc.tzinfo is None,
+                   'last_sync.date should be an offset-naive dt'):
+        last_sync_utc = UserTime(last_sync_utc).server_time().done()
```
Is `last_sync.date` server time?

https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/casexml/apps/phone/restore.py#L437

snopoke replied:
what's the scenario where this would happen?

dannyroberts replied:
Well, I think it wouldn't happen, and tracing the code it looks like it never happens, but the code I'm replacing had a provision for it having a tzinfo, so this is my no-risk way of asserting it doesn't.

czue commented:
@dannyroberts fyi, i'm hoping to do a pass on this this afternoon.

czue commented on corehq/apps/commtrack/models.py:
```diff
@@ -780,10 +782,10 @@ class Meta:
+    requested_on = USecDateTimeProperty()
+    approved_on = USecDateTimeProperty()
+    fulfilled_on = USecDateTimeProperty()
+    received_on = USecDateTimeProperty()
```
yowsers i totally forgot about requisitions. I wonder if we should kill them. (emailed a few people to discuss)

czue commented:
:+1: from me generally. not sure what we should be worried about breaking on the QA side

dannyroberts commented:
@NoahCarnahan Inheriting from `USecDocument` affects the behavior of "dynamic properties", i.e. properties that are not in the schema. Since form and case data is not defined in the schema, these are affected by that. Since datetimes are serialized to json as strings, if you're given a string and it looks like a datetime but it's not part of a schema, couchdbkit (and jsonobject after it) assumes it is a datetime. Using `USecDocument` changes the definition of "looks like a datetime" by changing the regex and then uses `USecDateTimeProperty` instead of `DateTimeProperty` to govern the semantics of it once it's been determined a string is a serialized date.

dannyroberts commented:
@czue part of my impetus for https://github.com/dimagi/commcare-hq/pull/6213 was to have an easy function to use instead of all the places that currently do a `strptime('%Y-%m-%dT%H:%M:%SZ')`. Doing that will, I imagine, get rid of the vast majority of places where this could go wrong.

NoahCarnahan commented:
Ah good to know thanks.

czue commented:
cool. @dannyroberts I'm signed off from a code perspective. defer to your judgment on QA or whether to pull the trigger.

dannyroberts committed:
Merge branch 'master' into microseconds

Conflicts:
	corehq/ex-submodules/casexml/apps/phone/caselogic.py

dannyroberts committed:
fix manual string parsing to use iso_string_to_datetime where appropriate

dannyroberts committed:
Merge branch 'master' into microseconds

Conflicts:
	corehq/apps/domain/models.py

dannyroberts commented:
Pulled out some more innocuous changes and put them here: https://github.com/dimagi/commcare-hq/pull/6235 to be merged faster and not create one giant diff. Once that's merged I have a few more changes building up on that that I'll add to this branch

dannyroberts committed:
Merge branch 'master' into microseconds

Conflicts:
	custom/_legacy/pact/reports/chw_schedule.py

dimagimon commented on corehq/apps/hqadmin/views.py:
```diff
@@ -469,7 +470,7 @@ def noneulized_users(request, template="hqadmin/noneulized_users.html"):
         reduce=False,
         include_docs=True,
-        startkey =["WebUser", days_ago.strftime("%Y-%m-%dT%H:%M:%SZ")],
+        startkey =["WebUser", datetime_to_iso_string(days_ago)],
```
E251 unexpected spaces around keyword / parameter equals

dannyroberts committed:
Merge branch 'master' into microseconds

Conflicts:
	corehq/apps/domain/models.py
